### PR TITLE
Add Smith-Waterman native bindings

### PR DESCRIPTION
**Pull request by gspowley**
Add native bindings for Smith-Waterman.

### CODE REVIEW

File: src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignmentResult.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.broadinstitute.gatk.nativebindings.smithwaterman;
+
+/**
+ * Struct used to hold results of the Smith-Waterman alignment result
+ */
+public class SWAlignmentResult {
+    // CIGAR string of the alignment
+    public String cigar;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignmentResult.java:8]: make these final and initialize in the constructor

File: src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeBinding.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     * Perform a Smith-Waterman alignment and return the result
+     *
+     * @param ref reference sequence
+     * @param alt alternate sequence
+     * @return alignment result
+     */
+    SWAlignmentResult align(byte[] ref, byte[] alt);
+
+    /**
+     * Print final profiling information from native code and close any resources.
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeBinding.java:35]: this should just close the resources

File: src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeArguments.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+         */
+        IGNORE
+    }
+    public OverhangStrategy strategy;
+
+    /**
+     * Smith-Waterman alignment parameters:
+     * match value, and mismatch, gap open and gap extension penalties
+     */
+    public int w_match;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeArguments.java:41]: these should be final and initialized in a constructor

**Review by lbergelson — CHANGES_REQUESTED** [review]

File: src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeBinding.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     */
+    void done();
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeBinding.java:37]: if we don't need this now, lets remove it and add it back if we do

File: src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeArguments.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    public int w_mismatch;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeArguments.java:42]: We should enforce any constraints on these in the constructor.

Comment (pnvaidya) [reply src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeArguments.java:42]: I tried to add constraints but I cannot use the Utils class which original code uses. Wondering if we can enforce constraints in base class in GATK as opposed to native bindings? thoughts?

Comment (lbergelson) [reply src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeArguments.java:42]: @pnvaidya Just use if statements and throw IllegalArgumentException.  That's all the validation methods do.

Comment (pnvaidya) [reply src/main/java/org/broadinstitute/gatk/nativebindings/smithwaterman/SWAlignerNativeArguments.java:42]: Ok. Done.

**Comment by lbergelson** [comment]
closing since these changes were completed in #11

**Pull request closed without merging**